crates/runtime/src/memory.rs

```diff
@@ -374,7 +442,131 @@ impl RuntimeLinearMemory for ExternalMemory {
         }
     }
 
-    #[cfg(feature = "pooling-allocator")]
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+/// For shared memory (and only for shared memory), this lock-version restricts
+/// access when growing the memory or checking its size. This is to conform with
+/// the [thread proposal]: "When `IsSharedArrayBuffer(...)` is true, the return
+/// value should be the result of an atomic read-modify-write of the new size to
+/// the internal `length` slot."
+///
+/// [thread proposal]:
+///     https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#webassemblymemoryprototypegrow
+#[derive(Clone)]
+pub struct SharedMemory(Arc<RwLock<SharedMemoryInner>>);
```

Comment:
Yeah, the `RwLock` could be considered overkill here. Looking at https://github.com/WebAssembly/threads/issues/26 shows some different options for implementing the locking around this. Initially I was considering locking everything (even `memory.size` through a host call) but since the addition of `AtomicUsize` I migrated towards locking `memory.grow` and allowing atomic access to the length. The read locks used in this struct could eventually be replaced by atomic loads and this `RwLock` with a `Mutex` but at the moment the extra safety seems fine and this type of refactor could be resolved in the tracking issue.

Reply:
No, it is possible that multiple threads could be asking for `byte_size()` and this allows those calls to be concurrent. Eventually we could move to using the `current_length: AtomicUsize` exclusively for `byte_size` but now that I think of it `maximum_byte_size` and `needs_init` still only need read access, not a full lock.